lib/auth/permissions.go

```diff
@@ -146,6 +146,26 @@ Loop:
 	return lockTargets
 }
 
+// UseSearchAsRoles replaces the checker on the current Context with the set of
+// roles the user is allowed to search as.
+func (c *Context) UseSearchAsRoles(access services.RoleGetter) error {
+	searchAsRoleNames := c.Checker.GetSearchAsRoles()
+	newRoleSet, err := services.FetchRoles(searchAsRoleNames, access, c.User.GetTraits())
+	if err != nil {
+		trace.Wrap(err)
+	}
+	if _, ok := c.Checker.(LocalUserRoleSet); ok {
+		c.Checker = LocalUserRoleSet{newRoleSet}
```

Comment:
Good question, I think either would work.

So far I went with overriding the roles here so that the results wouldn't be polluted with all the nodes they already have access to. But I'm not actually filtering those nodes out in case both rolesets would grant access, so it's not perfect.

But I'm starting to second-guess myself, I think we actually need to merge them for one case that would otherwise break. If you make a search-based access request you will lose access to all nodes except the ones in the request. If you need to access both a node you already have access to and one which you otherwise wouldn't at the same time, you may need to search for and request access to both.

For now I'll switch this to merge all the roles.